File: bridge/bindings/jsc/KOM/history.cc
/*
 * Copyright (C) 2021 Alibaba Inc. All rights reserved.
 * Author: Kraken Team.
 */

#include "history.h"
#include "dart_methods.h"
#include "bindings/jsc/KOM/window.h"
#include "bindings/jsc/foundation.h"

namespace kraken::binding::jsc {

std::stack<HistoryItem> JSHistory::m_previous_stack {};

std::stack<HistoryItem> JSHistory::m_next_stack {};

JSHistory::~JSHistory() {
}

JSValueRef JSHistory::getProperty(std::string &name, JSValueRef *exception) {
  if (name == "length") {
    return JSValueMakeNumber(context->context(), m_previous_stack.size() + m_next_stack.size());
  } else if (name == "state") {
    HistoryItem& history = m_previous_stack.top();
    if (history.state == nullptr) {
      return nullptr;
    } else {
      return JSValueMakeFromJSONString(ctx, history.state);
    }
  }

  return HostObject::getProperty(name, exception);
}

JSValueRef JSHistory::back(JSContextRef ctx, JSObjectRef function, JSObjectRef thisObject, size_t argumentCount,
                              const JSValueRef *arguments, JSValueRef *exception) {
  auto history = reinterpret_cast<JSHistory *>(JSObjectGetPrivate(function));

  if (m_previous_stack.size() > 1) {
    HistoryItem& currentItem = m_previous_stack.top();
    m_previous_stack.pop();
    m_next_stack.push(currentItem);

    history->goTo(m_previous_stack.top());

    JSStringRef state = m_previous_stack.top().state;
    history->dispatch(ctx, state, exception);
  }

  return nullptr;
}

JSValueRef JSHistory::forward(JSContextRef ctx, JSObjectRef function, JSObjectRef thisObject, size_t argumentCount,
                   const JSValueRef *arguments, JSValueRef *exception) {
  auto history = reinterpret_cast<JSHistory *>(JSObjectGetPrivate(function));

  if (!m_next_stack.empty()) {
    HistoryItem& currentItem = m_next_stack.top();
    m_next_stack.pop();
    m_previous_stack.push(currentItem);

    history->goTo(currentItem);

    JSStringRef &state = currentItem.state;
    history->dispatch(ctx, state, exception);
  }

  return nullptr;
}

JSValueRef JSHistory::go(JSContextRef ctx, JSObjectRef function, JSObjectRef thisObject, size_t argumentCount,
                              const JSValueRef *arguments, JSValueRef *exception) {
  JSValueRef value = arguments[0];
  int num = 0;
  if (JSValueIsString(ctx, value) || JSValueIsNumber(ctx, value)) {
    num = JSValueToNumber(ctx, value, exception);
  }

  auto history = reinterpret_cast<JSHistory *>(JSObjectGetPrivate(function));

  if (num >= 0) {
    if (m_next_stack.size() < num) {
      return nullptr;
    }

    for (int i = 0; i < num; i++) {
      HistoryItem& currentItem = m_next_stack.top();
      m_next_stack.pop();
      m_previous_stack.push(currentItem);
    }
  } else {
    if ((m_previous_stack.size() - 1) < abs(num)) {
      return nullptr;
    }

    for (int i = 0; i < num; i++) {
      HistoryItem& currentItem = m_previous_stack.top();
      m_previous_stack.pop();
      m_next_stack.push(currentItem);
    }
  }

  history->goTo(m_previous_stack.top());

  JSStringRef state = m_previous_stack.top().state;
  history->dispatch(ctx, state, exception);

  return nullptr;
}

void JSHistory::goTo(HistoryItem &historyItem) {
  if (!historyItem.needJump) return;

  NativeString *moduleName = stringRefToNativeString(JSStringCreateWithUTF8CString("Navigation"));
  NativeString *method = stringRefToNativeString(JSStringCreateWithUTF8CString("goTo"));
  NativeString *params = stringRefToNativeString(JSValueCreateJSONString(ctx, JSValueMakeString(ctx, historyItem.href), 0, nullptr));

  getDartMethod()->invokeModule(nullptr, context->getContextId(), moduleName, method, params,
                                handleInvokeModuleUnexpectedCallback);
}

JSStringRef JSHistory::getHref() {
  if (m_previous_stack.empty()) {
    return JSStringCreateWithUTF8CString("");
  }
  return m_previous_stack.top().href;
}

void JSHistory::dispatch(JSContextRef ctx, JSStringRef state, JSValueRef *exception) {
  JSStringHolder windowKeyHolder = JSStringHolder(context, "window");
  JSValueRef windowValue = JSObjectGetProperty(ctx, context->global(), windowKeyHolder.getString(), nullptr);
  JSObjectRef windowObject = JSValueToObject(context->context(), windowValue, nullptr);

  auto window = static_cast<WindowInstance *>(JSObjectGetPrivate(windowObject));

  std::string &&str = "popstate";
  auto nativeEvent = new NativeEvent(stringToNativeString(str));
  auto nativePopStateEvent = new NativePopStateEvent(nativeEvent);
  if (state != nullptr) {
    nativePopStateEvent->state = JSValueMakeFromJSONString(ctx, state);
  }

  std::string eventType = "popstate";
  auto event = JSEvent::buildEventInstance(eventType, context, nativePopStateEvent, false);

  window->dispatchEvent(event);
}

void JSHistory::addItem(HistoryItem &historyItem) {
  if (!m_previous_stack.empty() && historyItem.href == m_previous_stack.top().href) return;

  m_previous_stack.push(historyItem);

  // Clear.
  while(!m_next_stack.empty()) {
    m_next_stack.pop();
  }
}

JSValueRef JSHistory::pushState(JSContextRef ctx, JSObjectRef function, JSObjectRef thisObject, size_t argumentCount,
                               const JSValueRef *arguments, JSValueRef *exception) {
  if (argumentCount < 2) {
    throwJSError(ctx,
                 ("Failed to execute 'pushState' on 'History': 2 arguments required, but only " +
                  std::to_string(argumentCount) + " present")
                   .c_str(),
                 exception);
    return nullptr;
  }

  JSValueRef state = arguments[0];
  // https://github.com/whatwg/hpushStatetml/issues/2174
  JSValueRef title = arguments[1];
  JSValueRef url = arguments[2];
  JSStringRef urlRef = JSValueToStringCopy(ctx, url, exception);

  std::string strUrl = JSStringToStdString(urlRef);
  std::string strCurrentUrl = JSStringToStdString(m_previous_stack.top().href);

  Uri uri = Uri::Parse(strUrl);
  Uri currentUri = Uri::Parse(strCurrentUrl);

  if (uri.Host != "" && uri.Host != currentUri.Host) {
    throwJSError(ctx,
                 ("Failed to execute 'pushState' on 'History': A history state object with URL " + strUrl +
                  " cannot be created in a document with origin " + currentUri.Host + "  and URL " + strCurrentUrl + ".").c_str(),
                 exception);
  }

  JSStringRef jsonState = JSValueCreateJSONString(ctx, state, 0, exception);

  if (uri.Host == "" && uri.Protocol == "") {
    // Relative path.
    uri.Host = currentUri.Host;
    uri.Port = currentUri.Port;
    uri.Protocol = currentUri.Protocol;
  }

  HistoryItem history = { JSStringCreateWithUTF8CString(Uri::toString(uri).c_str()), jsonState, false };
  addItem(history);

  return nullptr;
}

JSValueRef JSHistory::replaceState(JSContextRef ctx, JSObjectRef function, JSObjectRef thisObject, size_t argumentCount,
                                   const JSValueRef *arguments, JSValueRef *exception) {
  if (argumentCount < 2) {
    throwJSError(ctx,
                 ("Failed to execute 'pushState' on 'History': 2 arguments required, but only " +
                  std::to_string(argumentCount) + " present")
                   .c_str(),
                 exception);
    return nullptr;
  }

  JSValueRef state = arguments[0];
  // https://github.com/whatwg/html/issues/2174
  JSValueRef title = arguments[1];
  JSValueRef url = arguments[2];
  JSStringRef urlRef = JSValueToStringCopy(ctx, url, exception);

  std::string strUrl = JSStringToStdString(urlRef);
  std::string strCurrentUrl = JSStringToStdString(m_previous_stack.top().href);

  Uri uri = Uri::Parse(strUrl);
  Uri currentUri = Uri::Parse(strCurrentUrl);

  if (uri.Host != "" && uri.Host != currentUri.Host) {
    throwJSError(ctx,
                 ("Failed to execute 'replaceState' on 'History': A history state object with URL " + strUrl +
                  " cannot be created in a document with origin " + currentUri.Host + "  and URL " + strCurrentUrl + ".").c_str(),
                 exception);
  }

  if (uri.Host == "" && uri.Protocol == "") {
    // Relative path.
    uri.Host = currentUri.Host;
    uri.Port = currentUri.Port;
    uri.Protocol = currentUri.Protocol;
  }

  JSStringRef jsonState = JSValueCreateJSONString(ctx, state, 0, exception);
  HistoryItem history = { JSStringCreateWithUTF8CString(Uri::toString(uri).c_str()), jsonState, false };

  m_previous_stack.pop();
  
  m_previous_stack.push(history);

  return nullptr;
}

void bindHistory(std::unique_ptr<JSContext> &context) {
  JSStringHolder windowKeyHolder = JSStringHolder(context.get(), "window");
  JSValueRef windowValue = JSObjectGetProperty(context->context(), context->global(), windowKeyHolder.getString(), nullptr);
  JSObjectRef windowObject = JSValueToObject(context->context(), windowValue, nullptr);
  auto window = static_cast<WindowInstance *>(JSObjectGetPrivate(windowObject));

  JSC_GLOBAL_SET_PROPERTY(context, "history", window->history_->jsObject);
}

} // namespace kraken::binding::jsc

File: webf/lib/src/module/history.dart
/*
 * Copyright (C) 2019-2022 The Kraken authors. All rights reserved.
 * Copyright (C) 2022-present The WebF authors. All rights reserved.
 */
import 'dart:async';
import 'dart:collection';
import 'dart:convert';

import 'package:webf/dom.dart';
import 'package:webf/webf.dart';

class HistoryItem {
  HistoryItem(this.bundle, this.state, this.needJump);
  final WebFBundle bundle;
  final dynamic state;
  final bool needJump;
}

class HistoryModule extends BaseModule {
  @override
  String get name => 'History';

  HistoryModule(ModuleManager? moduleManager) : super(moduleManager);

  Queue<HistoryItem> get _previousStack => moduleManager!.controller.previousHistoryStack;
  Queue<HistoryItem> get _nextStack => moduleManager!.controller.nextHistoryStack;

  WebFBundle? get stackTop {
    if (_previousStack.isEmpty) {
      return null;
    } else {
      return _previousStack.first.bundle;
    }
  }

  void add(WebFBundle bundle) {
    HistoryItem history = HistoryItem(bundle, null, true);
    _addItem(history);
  }

  void _addItem(HistoryItem historyItem) {
    if (_previousStack.isNotEmpty && historyItem.bundle.url == _previousStack.first.bundle.url) return;

    _previousStack.addFirst(historyItem);

    // Clear.
    while (_nextStack.isNotEmpty) {
      _nextStack.removeFirst();
    }
  }

  void _back() async {
    if (_previousStack.length > 1) {
      HistoryItem currentItem = _previousStack.first;
      _previousStack.removeFirst();
      _nextStack.addFirst(currentItem);

      await _goTo(_previousStack.first.bundle.url);

      dynamic state = _previousStack.first.state;
      _dispatchPopStateEvent(state);
    }
  }

  void _forward() {
    if (_nextStack.isNotEmpty) {
      HistoryItem currentItem = _nextStack.first;
      _nextStack.removeFirst();
      _previousStack.addFirst(currentItem);

      _goTo(currentItem.bundle.url);
      _dispatchPopStateEvent(currentItem.state);
    }
  }

  void _go(num? num) {
    num ??= 0;
    if (num >= 0) {
      if (_nextStack.length < num) {
        return;
      }

      for (int i = 0; i < num; i++) {
        HistoryItem currentItem = _nextStack.first;
        _nextStack.removeFirst();
        _previousStack.addFirst(currentItem);
      }
    } else {
      if (_previousStack.length - 1 < num.abs()) {
        return;
      }

      for (int i = 0; i < num.abs(); i++) {
        HistoryItem currentItem = _previousStack.first;
        _previousStack.removeFirst();
        _nextStack.addFirst(currentItem);
      }
    }

    _goTo(_previousStack.first.bundle.url);
    _dispatchPopStateEvent(_previousStack.first.state);
  }

  Future<void> _goTo(String targetUrl) async {
    NavigationModule navigationModule = moduleManager!.getModule<NavigationModule>('Navigation')!;
    await navigationModule.goTo(targetUrl);
  }

  void _dispatchPopStateEvent(state) {
    PopStateEventInit init = PopStateEventInit(state);
    PopStateEvent popStateEvent = PopStateEvent(init);
    moduleManager!.controller.view.window.dispatchEvent(popStateEvent);
  }

  void _pushState(List<dynamic> params) {
    WebFController controller = moduleManager!.controller;
    dynamic state = params[0];
    String? url = null;

    if (params[2] != null) {
      url = params[2];

      String currentUrl = _previousStack.first.bundle.url;
      Uri currentUri = Uri.parse(currentUrl);

      Uri uri = Uri.parse(url!);
      uri = controller.uriParser!.resolve(Uri.parse(controller.url), uri);

      if (uri.host.isNotEmpty && uri.host != currentUri.host) {
        print('Failed to execute \'pushState\' on \'History\': '
            'A history state object with URL $url cannot be created in a document with origin ${uri.host} and URL ${currentUri.host}. "');
        return;
      }

      WebFBundle bundle = WebFBundle.fromUrl(uri.toString());
      HistoryItem history = HistoryItem(bundle, state, false);
      _addItem(history);
    }
  }

  void _replaceState(List<dynamic> params) {
    WebFController controller = moduleManager!.controller;
    dynamic state = params[0];
    String? url = null;

    if (params[2] != null) {
      url = params[2];

      String currentUrl = _previousStack.first.bundle.url;
      Uri currentUri = Uri.parse(currentUrl);

      Uri uri = Uri.parse(url!);
      uri = controller.uriParser!.resolve(Uri.parse(controller.url), uri);

      if (uri.host.isNotEmpty && uri.host != currentUri.host) {
        print('Failed to execute \'pushState\' on \'History\': '
            'A history state object with URL $url cannot be created in a document with origin ${uri.host} and URL ${currentUri.host}. "');
        return;
      }

      WebFBundle bundle = WebFBundle.fromUrl(uri.toString());
      HistoryItem history = HistoryItem(bundle, state, false);

      _previousStack.removeFirst();
      _previousStack.addFirst(history);
    }
  }

  @override
  String invoke(String method, params, InvokeModuleCallback callback) {
    switch (method) {
      case 'length':
        return (_previousStack.length + _nextStack.length).toString();
      case 'state':
        if (_previousStack.isEmpty) {
          return jsonEncode(null);
        }
        HistoryItem history = _previousStack.first;
        return jsonEncode(history.state);
      case 'back':
        _back();
        break;
      case 'forward':
        _forward();
        break;
      case 'pushState':
        _pushState(params);
        break;
      case 'replaceState':
        _replaceState(params);
        break;
      case 'go':
        _go(params);
        break;
    }
    return EMPTY_STRING;
  }

  @override
  void dispose() {}
}
